### ISSUE TITLE
GDNative memory leak upon passing any Array

### ISSUE THREAD

**Issue by id01** [godotengine/godot#29391]
Similar to issue #19676

**Godot version:**

Still active in commit e9be875 of Godot, 0220045 of godot-cpp.

**OS/device including version:**
Arch Linux x64, with proprietary NVIDIA drivers.

**Issue description:**
Passing any array to a GDNative function causes a memory leak, as the array never gets freed.

The Array class causes a constant increase in static memory according to the debugger.

The PoolArray classes cause a constant increase in dynamic memory according to the debugger, and eventually causes a stack overflow after enough function calls (which is static memory; are these two stats switched around?).

**Steps to reproduce:**
```
void MyClass::memory_leak(Array anythingNotEmpty) {
  // Do nothing
}
```

Open the debugger in the editor when calling this function with an array of size >0. Go to Monitors->Memory->Static/Dynamic. Watch as your memory slowly (or quickly, depending on the size of the array) leaks away into oblivion. Faster for larger arrays. Weirdly, passing in the array as a Variant fixes the leak, but casting it back into an Array makes it leak again.

**Minimal reproduction project:**
Attached. I kept thinking my constant stack overflow segfaults were an issue with my C++ coding, until I switched from using PoolVector2Array to Array and my project ended up taking up a total of 8GB of RAM after a few minutes (all static).
[Godot_Memleak.zip](https://github.com/godotengine/godot/files/3244288/Godot_Memleak.zip)

**Comment by zhangshiqian1214** [comment]
I think this bug is same with #29128. Pass Array or PoolByteArray  to c++ will memory leak

**Comment by TheSHEEEP** [comment]
Oof, this is a bad one.
Is there any way around this, for the time being?

E.g., does it still leak when you pass the array as a Variant?

**Comment by TheSHEEEP** [comment]
This is really a bad one. Basically making use of Arrays impossible in GDNative.

Anyone having an idea what the cause of this could be?

**Comment by zhangshiqian1214** [comment]
I found a commit(https://github.com/GodotNativeTools/godot-cpp/pull/355/commits/95feb486c9968c4489f092df4232326d8ab7debf) in godot-cpp can fix this bug, The cast implicitly calls the C++ type's copy constructor. (You can find it in the gdb disassembly.) This increases the ref-count twice instead of just once through godot_variant_as_*. For PoolArrays, it caused an error message on exiting Godot

**Comment by akien-mga** [comment]
Fixed by GodotNativeTools/godot-cpp#355.

**Comment by Banderi** [comment]
Is there any fix for this in the C version?
